lib/Sema/CSSimplify.cpp

```diff
@@ -4768,6 +4800,145 @@ ConstraintSystem::simplifyKeyPathApplicationConstraint(
+      return markFailure();
+
+    // If we haven't seen a common result type yet, record what we found.
+    if (!commonResultType) {
+      commonResultType = choiceResultType;
+      return;
+    }
+
+    // If we already failed, we're done.
+    if (commonResultType->isNull())
+      return;
+
+    // If we found something different, fail.
+    if (!commonResultType.getValue()->isEqual(choiceResultType))
+      return markFailure();
+  };
+
+  // Consider each of the constraints in the disjunction.
+retry_after_fail:
+  bool skippedAnyConstraints = false;
+  bool labelMismatch = false;
+  auto filterResult =
+      filterDisjunctions(disjunction, /*restoreOnFail=*/shouldAttemptFixes(),
+                         [&](Constraint *constraint) {
+        // We must have bind-overload constraints.
+        // FIXME: This isn't entirely true.
+        if (constraint->getKind() != ConstraintKind::BindOverload) {
+          skippedAnyConstraints = true;
+          return true;
+        }
+
+        // We must be binding the type variable (or a type variable
+        // equivalent to it).
+        auto boundTypeVar = constraint->getFirstType()->getAs<TypeVariableType>();
+        if (!boundTypeVar || getRepresentative(boundTypeVar) != fnTypeVar) {
+          skippedAnyConstraints = true;
+          return true;
+        }
+
+        auto choice = constraint->getOverloadChoice();
+
+        // Determine whether the argument labels we have conflict with those of
+        // this overload choice.
+        if (argumentLabels &&
+            !areConservativelyCompatibleArgumentLabels(
+                choice, argumentLabels->Labels,
+                argumentLabels->HasTrailingClosure)) {
+          labelMismatch = true;
+          return false;
+        }
+
+        // Determine the type that this choice will have.
+        Type choiceType =
+            getEffectiveOverloadType(choice, /*allowMembers=*/true,
+                                     constraint->getOverloadUseDC());
+        if (!choiceType) {
+          skippedAnyConstraints = true;
+          return true;
+        }
+
+        // If we have a function type, we can compute a common result type.
+        updateCommonResultType(choiceType);
+        return true;
+      });
+
+  switch (filterResult) {
+  case SolutionKind::Error:
+    if (labelMismatch && shouldAttemptFixes()) {
+      argumentLabels = None;
+      goto retry_after_fail;
+    }
+
+    return Type();
+
+  case SolutionKind::Solved:
+    // We should now have a type for the one remaining overload.
+    fnType = getFixedTypeRecursive(fnType, /*wantRValue=*/true);
+    break;
+
+  case SolutionKind::Unsolved:
+    break;
+  }
+
+
+  // If there was a constraint that we couldn't reason about, don't use the
+  // results of any common-type computations.
+  if (skippedAnyConstraints)
+    return fnType;
+
+  // If we have a common result type, bind the expected result type to it.
+  if (commonResultType && *commonResultType) {
```

Comment:
This seems a bit weird to me - should setting an optional with empty/invalid type be avoided?

Reply:
I need three states: not computed, computed, and conflicting. I could use a `Type` with a different sentinel (e.g., `ErrorType`), I guess.

Reply:
I think it might make sense to use Type here for this, since this check means - computed but empty/invalid)?

Reply:
Without the optional, it'll mean doing a `commonResultType && !commonResultType->is<ErrorType>()` here, which is only slightly better.

Reply:
I don’t have much preference here, but maybe UresolvedType would be more appropriate to indicate conflict. You can leave it as optional here, it just seems like a quirk of C++ where you can have optional of null pointer...

Reply:
`UnresolvedType` is something I'd like us to remove, so I'd rather not create a dependency on it. I'll go with `ErrorType`.

Reply:
Sure!